File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> answer; 
    void dfs(vector<int>& candidates, vector<int>& container, int& target, int index){
        if(target <= 0){
            if(target == 0) answer.push_back(container); 
            return;  
        }

        for(int i = index; i < candidates.size(); i++){
            container.push_back(candidates[i]); 
            target -= candidates[i]; 
            dfs(candidates, container, target, i); 
            container.pop_back();
            target += candidates[i]; 
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> container; 
        
        dfs(candidates,container,target,0); 

        return answer; 
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> answer = new ArrayList<>(); 
        List<Integer> container = new ArrayList<>(); 
        dfs(candidates, target, answer, container, 0); 
        return answer; 
    }
    void dfs(int[] candidates, int target, List<List<Integer>> answer, List<Integer> container, int index){
        if(target <= 0){
            if(target == 0) answer.add(new ArrayList<>(container)); 
            return; 
        }

        for(int i = index; i < candidates.length; i++){
            target -= candidates[i]; 
            container.add(candidates[i]); 
            dfs(candidates, target, answer, container, i); 
            target += candidates[i]; 
            container.removeLast();  
        }
    }
}